controlplane/kubeadm/internal/controllers/upgrade_test.go

```diff
@@ -115,17 +172,24 @@ func TestKubeadmControlPlaneReconciler_RolloutStrategy_ScaleUp(t *testing.T) {
 	}
 	controlPlane.Machines = collections.FromMachineList(bothMachines)
 
+	machinesRequireUpgrade := collections.Machines{}
+	for i := range bothMachines.Items {
+		if bothMachines.Items[i].Spec.Version != nil && *bothMachines.Items[i].Spec.Version != UpdatedVersion {
+			machinesRequireUpgrade[bothMachines.Items[i].Name] = &bothMachines.Items[i]
+		}
+	}
+
 	// run upgrade the second time, expect we scale down
-	result, err = r.upgradeControlPlane(ctx, cluster, kcp, controlPlane, controlPlane.Machines)
+	result, err = r.upgradeControlPlane(ctx, cluster, kcp, controlPlane, machinesRequireUpgrade)
 	g.Expect(err).To(BeNil())
 	g.Expect(result).To(Equal(ctrl.Result{Requeue: true}))
 	finalMachine := &clusterv1.MachineList{}
-	g.Expect(fakeClient.List(ctx, finalMachine, client.InNamespace(cluster.Namespace))).To(Succeed())
-	g.Expect(finalMachine.Items).To(HaveLen(1))
-
-	// assert that the deleted machine is the oldest, initial machine
-	g.Expect(finalMachine.Items[0].Name).ToNot(Equal(initialMachine.Items[0].Name))
-	g.Expect(finalMachine.Items[0].CreationTimestamp.Time).To(BeTemporally(">", initialMachine.Items[0].CreationTimestamp.Time))
+	g.Eventually(func(g Gomega) {
+		g.Expect(env.List(ctx, finalMachine, client.InNamespace(cluster.Namespace))).To(Succeed())
+		g.Expect(finalMachine.Items).To(HaveLen(1))
+		// assert that the deleted machine is the oldest, initial machine
```

Comment:
The test still checks that the old machine is deleted and the new machine is kept around by checking the name.

I had to drop the creationTimestamp based additional check because now that we switched to envtest for this test the real apiserver apparently does not have enough resolution to time for this test. Because of how quickly the test runs the creationTimestamps of the new and old machines are equal in most runs. 
Note that since we are verifying that the final list of machines is of length 1 and that the only machine is not the old machine we can guarantee that the old machine is the one that got deleted. Therefore it is safe to drop the creationTimestamp check here.

Reply:
Sounds good. Thx for the explanation.

I thought either we should assert it again or change the comment. Changing the comment makes sense!